### PR TITLE
[LayoutAnimation] RCTAnimationTypeKeyboard

### PR DESCRIPTION
**Pull request by vishnevskiy**
This adds the Keyboard animation type for when you want to animate UI based on the keyboard appearing/disappearing.

### CODE REVIEW

**Comment by ide** [comment]
Is the animation curve always the same? Hiding/showing the autosuggest bar on iOS 8 is also considered a keyboard animation and may use a different curve. For robustness and forwards-compatibility it might be better to expose all of the built-in animation curves, and have the app specify which one to use.

**Comment by vishnevskiy** [comment]
Honestly I am not sure if there are other ones. This is the only one I have ever used. Would be awesome if someone else could comment on this issue with more experience with the matter.

**Comment by a2** [comment]
This seems like a bad idea considering it's undocumented and likely to change at any time. I vote to close this. @nicklockwood, thoughts?

**Comment by vishnevskiy** [comment]
@a2 Without such an animation available how do you plan to make it possible to animate at the same rate as the keyboard? Normally people just take the animation curve from the Keyboard events.

**Comment by ide** [comment]
@vishnevskiy, RN apps should do the same and use the animation curve specified by the keyboard events.

**Comment by vishnevskiy** [comment]
I guess the problem is the animation currently takes an enum, prob should be extended to support an arbitrary value as well. That would be a better approach so feel free to close this :)

**Comment by a2** [comment]
If we can support it, I would prefer to use `inputAccessoryView` to achieve this instead of a keyboard animation type.

**Comment by josebalius** [comment]
Guys.....so what would be the proper way to use this? I am not sure if this is what I am looking for, but essentially I want a field to animate up with the same speed as the keyboard....is this possible with these commits?

File: Libraries/WebSocket/RCTWebSocketManager.m [hunk 1]
```diff
@@ -107,7 +107,7 @@ - (void)webSocket:(RCTSRWebSocket *)webSocket didCloseWithCode:(NSInteger)code
 {
   [_bridge.eventDispatcher sendDeviceEventWithName:@"websocketClosed" body:@{
     @"code": @(code),
-    @"reason": reason,
+    @"reason": reason ? reason : [NSNull null],
```
Comment (a2) [inline Libraries/WebSocket/RCTWebSocketManager.m]: This should be in a separate diff. It's unrelated to keyboard changes.

File: Libraries/Animation/LayoutAnimation.js [hunk 2]
```diff
@@ -23,6 +23,7 @@ var TypesEnum = {
   easeInEaseOut: true,
   easeIn: true,
   easeOut: true,
+  keyboard: true
```
Comment (a2) [inline Libraries/Animation/LayoutAnimation.js]: Please add trailing comma.

**Comment by a2** [comment]
@vishnevskiy Looks good if we can fix those couple nits.

**Comment by vishnevskiy** [comment]
@a2 Sorry about that no idea how those diffs got into this pull request. Will fix.

**Comment by sahrens** [comment]
@a2 I'm not sure we really want this - inputAccessoryView is probably the better way to go, but obviously more work than this...but putting in your queue to either ship it or close.

**Comment by a2** [comment]
I agree that I don't think this is the best way to do. @nicklockwood, what do you think? We spoke offline about this but I don't remember your opinion.

**Comment by nicklockwood** [comment]
My view is essentially the same as @sahrens's - the right solution is to implement inputAccessoryView, but that's more difficult, and this might be a reasonable stopgap solution until we can do that.

**Comment by a2** [comment]
@nicklockwood I agree with what you're saying. Do we want to merge this for now or should someone write `inputAccessoryView` support?

**Comment by nicklockwood** [comment]
@a2 both?

**Comment by nicklockwood** [comment]
If I understand correctly, @vjeux's animation system will make the existing LayoutAnimation API obsolete, so if the whole system is a temporary workaround anyway, I don't see any harm in adding some additional workarounds to it.

**Comment by nicklockwood** [comment]
I have some ideas for how inputAccessoryView can be implemented. If somebody has some time to look at building it, I'm happy to discuss solutions.

**Comment by sahrens** [comment]
Vjeux new animation stuff doesn't replace LayoutAnimation at all - they are
actually quite complimentary and both can do things the other cannot (for
example, animating flexWrap grid re-orderings).

File: React/Modules/RCTUIManager.m [hunk 3]
```diff
@@ -69,6 +69,9 @@ static UIViewAnimationCurve UIViewAnimationCurveFromRCTAnimationType(RCTAnimatio
       return UIViewAnimationCurveEaseOut;
     case RCTAnimationTypeEaseInEaseOut:
       return UIViewAnimationCurveEaseInOut;
+    case RCTAnimationTypeKeyboard:
+      // http://stackoverflow.com/questions/18870447/how-to-use-the-default-ios7-uianimation-curve
+      return (UIViewAnimationCurve) (7 << 16);
```
Comment (a2) [inline React/Modules/RCTUIManager.m]: This would be `(UIAnimationCurve)7` but `(UIViewAnimationOptions)(7 << 16)`.

**Comment by a2** [comment]
Merged internally.

**Pull request closed without merging**